### DISCUSSION TITLE
Symfony process output streaming stops abruptly

### DISCUSSION THREAD

**Discussion by carlalexander** [symfony/symfony#45817] [Q&A] [answered]
I've been trying to debug this weird issue with Symfony Process. The output streaming stops working while a long process is running. This is a simplified version of the code.

```php
$process = new Process(['long-running-process']);
$process->setTimeout(null);
$process->start(function ($type, $output) {
    fwrite(STDERR, 'Process: '.$output.PHP_EOL);
});
```

This works and then stops. I've been able to confirm that process still sends output because, if I start doing something like this, the output from the callback appears again.

```php
$process = new Process(['long-running-process']);
$process->setTimeout(null);
$process->start(function ($type, $output) {
    fwrite(STDERR, 'Process: '.$output.PHP_EOL);
});

do {
    $process->getIncrementalErrorOutput();
while ($process->isRunning());
```

I'm really confused by what would cause something like this to happen. I manually added some trace inside the component code. It seems like `readPipes` is being regularly called and then it stops. After that point, the only way `readPipes` gets triggered is via `getIncrementalErrorOutput`.

Anyone know what I could look into to try to figure out what's going on?

**Comment by lyrixx** [comment]
It's legit. `Process::start()` only starts the process. Then you have to wait for the process to run.
You can replace `start` by `run` if you want the process to finish

I often suggest to use `Process::mustRun()`, to get a nice exception is something goes wrong.

I invite you to read the PHP Doc, that, hopefully, explain everything: https://github.com/symfony/symfony/blob/7124c3a0f5f0f28d857dc9573348940542c9dee1/src/Symfony/Component/Process/Process.php#L273-L292

**Comment by carlalexander** [comment, marked as answer]
Thanks @lyrixx! I'd read all of this, but yeah, it didn't click.

The issue more specifically is I didn't have any code that triggered `updateStatus`. This is what causes `readPipes` to run. I was missing code that actually did that. Even in the example I gave, I could have done:

```php
do {
    sleep(1);
while ($process->isRunning());
```

`isRunning` triggers a `updateStatus` call so that's all I needed.

Thanks <3

**Reply by stof** [reply]
if you just want to wait until the process finishes running, using `$process->wait()` is simpler than building this loop yourselves

**Reply by carlalexander** [reply]
@stof I can't use `wait` because it's a daemon managed by Symfony that'll run for an indefinite amount of time. I just need to pipe what it's sending to `STDERR` to the application logs.
